edge/pkg/devicetwin/dtmanager/communicate_test.go

```diff
@@ -0,0 +1,319 @@
+			wantErr:  nil,
+		},
+		{
+			name:     "dealLifeCycleTest-CloudNotConnected",
+			context:  dtContext,
+			resource: "",
+			msg:      &model.Message{Content: model.CloudDisconnected},
+			want:     nil,
+			wantErr:  nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := dealLifeCycle(test.context, test.resource, test.msg)
+			if !reflect.DeepEqual(err, test.wantErr) {
+				t.Errorf("dealLifeCycle() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("dealLifeCycle() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestDealConfirm is function to test dealConfirm().
+func TestDealConfirm(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContext, _ := dtcontext.InitDTContext(mainContext)
+	tests := []struct {
+		name     string
+		context  *dtcontext.DTContext
+		resource string
+		msg      interface{}
+		want     interface{}
+		wantErr  error
+	}{
+		{
+			name:    "dealConfirmTest-WrongMsg",
+			context: dtContext,
+			msg:     "",
+			want:    nil,
+			wantErr: errors.New("CommModule deal confirm, type not correct"),
+		},
+		{
+			name:    "dealConfirmTest-CorrectMsg",
+			context: dtContext,
+			msg:     &model.Message{},
+			want:    nil,
+			wantErr: nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := dealConfirm(test.context, test.resource, test.msg)
+			if !reflect.DeepEqual(err, test.wantErr) {
+				t.Errorf("dealConfirm() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("dealConfirm() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestDetailRequest is function to test detailRequest().
+func TestDetailRequest(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContext, _ := dtcontext.InitDTContext(mainContext)
+	tests := []struct {
+		name    string
+		context *dtcontext.DTContext
+		msg     interface{}
+		want    interface{}
+		wantErr error
+	}{
+		{
+			name:    "detailRequestTest",
+			context: dtContext,
+			msg:     "",
+			wantErr: nil,
+			want:    nil,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			got, err := detailRequest(test.context, test.msg)
+			if !reflect.DeepEqual(err, test.wantErr) {
+				t.Errorf("detailRequest() error = %v, wantErr %v", err, test.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, test.want) {
+				t.Errorf("detailRequest() = %v, want %v", got, test.want)
+			}
+		})
+	}
+}
+
+// TestCheckConfirm is function to test checkConfirm().
+func TestCheckConfirm(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContext, _ := dtcontext.InitDTContext(mainContext)
+	dtContext.ConfirmMap.Store("key", &dttype.DTMessage{})
+	dtContext.ConfirmMap.Store("key", &dttype.DTMessage{Action: dtcommon.SendToCloud})
```

Comment:
we overwrite the same key ? how will it help test the behaviour of the confirm method ?

Reply:
Updated